pilot/pkg/xds/ads.go

```diff
@@ -386,6 +386,8 @@ func (s *DiscoveryServer) initializeProxy(con *Connection) error {
 }
 
 func (s *DiscoveryServer) computeProxyState(proxy *model.Proxy, request *model.PushRequest) {
+	proxy.Lock()
+	defer proxy.Unlock()
```

Comment:
What does it really protect? If it is to protect the whole proxy struct, should be called at more places

Reply:
There are like 6 fields in this function that are set - any one of them races with the debug handler that tries to read them

Reply:
There are still many `read` in debug.go that not protected, do they not need to protect?
```
			Labels:       c.proxy.Labels,
			Metadata:     c.proxy.Metadata,
			Locality:     c.proxy.Locality,
```

```
proxy.LastPushContext,
```

```
proxy.SidecarScope
```

Reply:
the debug code clones the proxy before accessing things so it's more racing during the clone than direct accesses to individual fields afaik

Reply:
ah, right